core/src/test/scala/org/apache/spark/scheduler/BarrierTaskContextSuite.scala

```diff
@@ -276,4 +276,20 @@ class BarrierTaskContextSuite extends SparkFunSuite with LocalSparkContext {
     initLocalClusterSparkContext()
     testBarrierTaskKilled(interruptOnKill = true)
   }
+
+  test("SPARK-31485: barrier stage should fail if only partial tasks are launched") {
+    initLocalClusterSparkContext(2)
+    val rdd0 = sc.parallelize(Seq(0, 1, 2, 3), 2)
+    val dep = new OneToOneDependency[Int](rdd0)
+    // set up a barrier stage with 2 tasks and both tasks prefer executor 0 (only 1 core) for
```

Comment:
this seems wrong to me, locality preference is causing failure?  I thought we should be looking at all available slots and only erroring if you didn't have enough.  Here you have 2 workers with 1 cores each so you should be able to fit 2 tasks on there eventually.

Reply:
> this seems wrong to me, locality preference is causing failure?

Yeah, I have to say so. We do have enough slots in this case but it doesn't guarantee that all tasks in a task sets can be scheduled in a single round resource offer because of delay scheduling. While partial scheduling is ok for normal task set, it's unacceptable for barrier task set.

Reply:
ok looking through some of the logic some more, I guess if total slots is less then it skips it higher up. This still seems very odd that you fail based on locality being set.  Wouldn't you just want to ignore locality in this case?   Dynamic allocation isn't on, if you have number slots = number of tasks you need to just schedule it.

Do we at least recommend people turning off locality when using barrier?  I recommend shutting off current to most people anyway on yarn because it can have other bad issues.  Do we have a jira for this issue, this seems like it could be very confusing to users.  The message says something about blacklisted, so if I'm going to try to debug to figure out why my stuff isn't scheduled, I think this would be very hard to figure out.

Reply:
Note my last comment was posted before I saw your response.  I get that you don't have enough slots here, but you are failing the task set, in the case of locality if you just wait you would likely get it.  Or like my last comment says we could recognize it, or in the very least add locality to the message for our failure.

Reply:
The `require(addressesWithDescs.size == taskSet.numTasks)` there is more like a safe check for the barrier task set and prevents any possible reasons from partial tasks launching. Obviously, delay scheduling is one of reasons we can tell. I agree to improve error message against delay scheduling if it's also the only reason.

Reply:
Also cc @jiangxb1987

Reply:
we can certainly do it separate from this since that appears to be existing behavior. I was surprised by it because I was thinking of other cases like blacklisting, which makes more sense to me

Reply:
`makes more sense`? Or, no? I think blacklist offers have already been filtered out at that point?

Reply:
@Ngone51 can you create a JIRA ticket? I think it's worth discussion to see if we should reject to schedule barrier tasks due to locality.

Reply:
Added https://issues.apache.org/jira/browse/SPARK-31509